shell/app/node_main.cc

```diff
@@ -116,28 +116,40 @@ int NodeMain(int argc, char* argv[]) {
     }
 
     node::LoadEnvironment(env);
+    v8::Isolate* isolate = env->isolate();
+
+    {
+      v8::SealHandleScope seal(isolate);
+      bool more;
+      do {
+        uv_run(env->event_loop(), UV_RUN_DEFAULT);
+
+        gin_env.platform()->DrainTasks(env->isolate());
 
-    bool more;
-    do {
-      more = uv_run(env->event_loop(), UV_RUN_ONCE);
-      gin_env.platform()->DrainTasks(env->isolate());
-      if (more == false) {
-        node::EmitBeforeExit(env);
+        more = uv_loop_alive(env->event_loop());
+        if (more && !env->is_stopping())
+          continue;
+
+        if (!uv_loop_alive(env->event_loop())) {
+          EmitBeforeExit(env);
+        }
 
         // Emit `beforeExit` if the loop became alive either after emitting
         // event, or after running some callbacks.
```

Comment:
should this comment be moved above the call to `EmitBeforeExit`?

Reply:
@nornagon i'm fairly impartial on this - i felt it better to directly emulate Node.js itself, and this is where the comment is in that codebase

Reply:
ah ok, didn't realise this was copy/pasted from node directly.